packages/opentelemetry-node-tracer/src/instrumentation/PluginLoader.ts

```diff
@@ -0,0 +1,149 @@
+}
+
+interface PluginNames {
+  [pluginName: string]: string;
+}
+
+interface PluginConfig {
+  // TODO: Consider to add configuration options
+  [pluginName: string]: boolean;
+}
+
+// tslint:disable-next-line:no-any
+export type PluginType = Array<Plugin<any>>;
+
+/**
+ * The PluginLoader class can load instrumentation plugins that
+ * use a patch mechanism to enable automatic tracing for
+ * specific target modules.
+ */
+export class PluginLoader {
+  /** A list of loaded plugins. */
+  private _plugins: PluginType = [];
+  /**
+   * A field that tracks whether the require-in-the-middle hook has been loaded
+   * for the first time, as well as whether the hook body is activated or not.
+   */
+  private _hookState = HookState.UNINITIALIZED;
+
+  /** Constructs a new PluginLoader instance. */
+  constructor(readonly tracer: Tracer, readonly logger: Logger) {}
+
+  /**
+   * Loads a list of plugins. Each plugin module should implement the core
+   * {@link Plugin} interface and export an instance named as 'plugin'. This
+   * function will attach a hook to be called the first time the module is
+   * loaded.
+   * @param pluginConfig an object whose keys are plugin names and whose
+   *     boolean values indicate whether to enable the plugin.
+   */
+  load(pluginConfig: PluginConfig): PluginLoader {
+    if (this._hookState === HookState.UNINITIALIZED) {
+      const plugins = Object.keys(pluginConfig).reduce(
+        (plugins: PluginNames, moduleName: string) => {
+          if (pluginConfig[moduleName]) {
+            plugins[moduleName] = utils.defaultPackageName(moduleName);
+          }
+          return plugins;
+        },
+        {} as PluginNames
+      );
+      const modulesToHook = Object.keys(plugins);
+      // Do not hook require when no module is provided. In this case it is
+      // not necessary. With skipping this step we lower our footprint in
+      // customer applications and require-in-the-middle won't show up in CPU
+      // frames.
+      if (modulesToHook.length === 0) {
+        this._hookState = HookState.DISABLED;
+        return this;
+      }
+
+      // Enable the require hook.
+      hook(modulesToHook, (exports, name, baseDir) => {
+        if (this._hookState !== HookState.ENABLED) return exports;
+
+        const moduleName = plugins[name];
+        // Get the module version.
+        const version = utils.getPackageVersion(this.logger, baseDir as string);
+        this.logger.info(
+          `PluginLoader#load: trying loading ${name}.${version}`
+        );
+
+        // @todo (issues/132): Check if version and supportedVersions are
+        // satisfied
+        if (!version) return exports;
+
+        this.logger.debug(
+          `PluginLoader#load: applying patch to ${name}@${version} using ${moduleName} module`
+        );
+
+        // Expecting a plugin from module;
+        try {
+          // tslint:disable-next-line:no-any
+          const plugin: Plugin<any> = require(moduleName).plugin;
```

Comment:
Seems like we are pretty much always doing `Plugin<any>`. What if we move the `any` to the Plugin type so that it because `class Plugin<T = any> {`?

Reply:
Ok, done.